File: Leetcode/Leetcode2/EvalRPN/EvalRPN.cpp
#include <vector>
#include <string>
#include <stack>

using namespace std;

class Solution {
private:
    int sToI(string s) {
        int ans = 0, k = 0;
        if (s[0] == '-')
            k = 1;
        for (int i = k; i < s.size(); i++)
            ans = ans * 10 + (s[i] - '0');
        return s[0] == '-' ? -ans : ans;
    }
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> s;
        for (auto& token : tokens)
        {
            if (token.size() > 1)
            {
                s.push(sToI(token));
                continue;
            }
            int num1, num2;
            switch (token[0])
            {
            case '+':
                num2 = s.top();
                s.pop();
                num1 = s.top();
                s.pop();
                s.push(num1 + num2);
                break;
            case '-':
                num2 = s.top();
                s.pop();
                num1 = s.top();
                s.pop();
                s.push(num1 - num2);
                break;
            case '*':
                num2 = s.top();
                s.pop();
                num1 = s.top();
                s.pop();
                s.push(num1 * num2);
                break;
            case '/':
                num2 = s.top();
                s.pop();
                num1 = s.top();
                s.pop();
                s.push(num1 / num2);
                break;
            default:
                s.push(sToI(token));
                break;
            }
        }
        return s.top();
    }
};

File: Leetcode/Leetcode9/EvalRPN.java
package Leetcode.Leetcode9;

import java.util.Stack;

public class EvalRPN {
    private boolean isInteger(String str) {
        if (str == null || str.isBlank()) return false;
        try {
            Integer.parseInt(str);
            return true;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    public int evalRPN(String[] tokens) {
        Stack<Integer> s1 = new Stack<>();
        for (String token : tokens) {
            if (isInteger(token)) {
                s1.push(Integer.parseInt(token));
            } else {
                int a = s1.pop(), b = s1.pop();
                switch (token) {
                    case "+":
                        s1.push(a + b);
                        break;
                    case "-":
                        s1.push(b - a);
                        break;
                    case "*":
                        s1.push(a * b);
                        break;
                    default:
                        s1.push(b / a);
                        break;
                }
            }
        }
        return s1.pop();
    }
}
